crates/server/src/store/file/authority.rs

```diff
@@ -54,6 +58,66 @@ impl FileAuthority {
         InMemoryAuthority::new(origin, records, zone_type, allow_axfr).map(Self)
     }
 
+    /// Read given file line by line and recursively invokes reader for
+    /// $INCLUDE directives
+    ///
+    /// TODO: it looks hacky as far we effectively duplicate parser's functionallity
+    /// (at least partially) and performing lexing twice.
+    /// Better solution requires us to change lexer to deal
+    /// with Lines-like iterator instead of String buf (or capability to combine a few
+    /// lexer instances into a single lexer).
+    ///
+    /// TODO: we should probably track depth of recursion here and set max level (or
+    /// even make it configurable, in practice it's hard to expect INCLUDES deeper than
+    /// 3-4 levels anyways).
+    ///
+    /// TODO: $INCLUDE could specify domain name -- to support on-flight swap for Origin
+    /// value we definitely need to rethink and rework loader/parser/lexer
+    fn read_file(zone_path: PathBuf, buf: &mut String) -> Result<(), String> {
+        let file = File::open(&zone_path)
+            .map_err(|e| format!("failed to read {}: {:?}", zone_path.display(), e))?;
+        let reader = BufReader::new(file);
+        for line in reader.lines() {
+            let content = line.unwrap();
+            let mut lexer = Lexer::new(&content);
+
+            match (lexer.next_token(), lexer.next_token()) {
+                (Ok(Some(Token::Include)), Ok(Some(Token::CharData(include_path)))) => {
+                    // RFC1035 (section 5) does not specify how filename for $INCLUDE
+                    // should be resolved into file path. The underlying code implements the
+                    // following:
+                    // * if the path is absolute (relies on Path::is_absolute), it uses normalized path
+                    // * otherwise, it joins the path with parent root of the current file
+                    //
+                    // TODO: Inlining files specified using non-relative path might potentially introduce
```

Comment:
I agree with the concern here, not sure if I have a good option, but we should think about it.

Reply:
I checked how other DNS servers deal with this situations. Notably,
* [PowerDNS](https://github.com/PowerDNS/pdns/blob/0789d16b108a4f4d57a5c3d743c7a489a4ce9f7e/pdns/zoneparser-tng.cc#L314-L315) allows absolute path without restrictions 

* [Unbound](https://github.com/NLnetLabs/unbound/blob/d000523b000c1899cb23673297d1aba550df08c8/services/authzone.c#L1497-L1501) checks that $INCLUDE stays within configurable `chroot` if the latest is specified 

@bluejekyll Leaving the final call on you ;)

Reply:
I think it's reasonable to allow any path (relative to the file including this, or fully specified). If folks want a stronger security stance, then using chroot is always an option.